lib/page-loader.js

```diff
@@ -0,0 +1,104 @@
+/* global window, document */
+import mitt from 'mitt'
+
+const webpackModule = module
+
+export default class PageLoader {
+  constructor (buildId) {
+    this.buildId = buildId
+    this.pageCache = {}
+    this.pageLoadedHandlers = {}
+    this.registerEvents = mitt()
+    this.loadingRoutes = {}
+  }
+
+  normalizeRoute (route) {
+    if (route[0] !== '/') {
+      throw new Error('Route name should start with a "/"')
+    }
+
+    return route.replace(/index$/, '')
+  }
+
+  loadPage (route) {
+    route = this.normalizeRoute(route)
+
+    const cachedPage = this.pageCache[route]
```

Comment:
It looks cache is not used when multiple concurrent requests happen for a route at a same time. maybe we can improve that later tho.